### PR TITLE
    fix(tianmu): Fixed implicit convert from number to varchar(#959)

### LINKED ISSUE #959: bug: Query result set error，varchar(255) The result of adding data is not queried
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/stoneatom/stonedb/blob/stonedb-5.7-dev/CONTRIBUTING.md).

### Please confirm if bug report does NOT exists already ?

- [X] I confirm there is no existing issue for this

### Describe the problem

```
CREATE TABLE device_delta_collect
(
    reportTime varchar(255) DEFAULT NULL,
    deviceId   varchar(255) DEFAULT NULL,
    multiple   varchar(255) DEFAULT NULL,
    deltaJgs   varchar(255) DEFAULT NULL
) ENGINE = tianmu;

insert into device_delta_collect (reportTime, deviceId, deltaJgs, multiple) values ('2022-08-20 22:33:58', '1547034559760195585', '1', '1');

select * from device_delta_collect a where a.deviceId = 1547034559760195587 limit 10;
Empty set (0.00 sec)
```

### Expected behavior

MySQL(innodb) results:
```
mysql> select * from device_delta_collect a where a.deviceId = 1547034559760195587 limit 10;
+---------------------+---------------------+----------+----------+
| reportTime          | deviceId            | multiple | deltaJgs |
+---------------------+---------------------+----------+----------+
| 2022-08-20 22:33:58 | 1547034559760195585 | 1        | 1        |
+---------------------+---------------------+----------+----------+
```

### How To Reproduce

```
CREATE TABLE device_delta_collect
(
    reportTime varchar(255) DEFAULT NULL,
    deviceId   varchar(255) DEFAULT NULL,
    multiple   varchar(255) DEFAULT NULL,
    deltaJgs   varchar(255) DEFAULT NULL
) ENGINE = tianmu;

insert into device_delta_collect (reportTime, deviceId, deltaJgs, multiple) values ('2022-08-20 22:33:58', '1547034559760195585', '1', '1');

select * from device_delta_collect a where a.deviceId = 1547034559760195587 limit 10;
```

### Environment

```
root@ub01:/stonedb57/install/bin# ./mysql --version
./mysql  Ver 14.14 Distrib 5.7.36-StoneDB, for Linux (x86_64) using  EditLine wrapper
```

### Are you interested in submitting a PR to solve the problem?

- [ ] Yes, I will!

### PR DESCRIPTION
**Pull request by adofsauron**
## Summary about this PR

Issue Number: close #959 

## Tests Check List

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

## Changelog

- [ ] New Feature
- [x] Bug Fix
- [ ] Performance Improvement
- [ ] Build/Testing/CI/CD
- [ ] Documentation
- [ ] Not for changelog (changelog entry is not required)

## Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features

### CODE REVIEW

File: storage/tianmu/types/tianmu_data_types.h [hunk 1]
```diff
@@ -490,6 +498,21 @@ static int inline CollationStrCmp(DTCollation coll, const BString &s1, const BSt
                                          s2.len_, 0);
 }
 
+static int inline CollationRealCmp(DTCollation coll, const BString &s1, const BString &s2) {
+  int not_used;
```
Comment (lujiashun) [inline storage/tianmu/types/tianmu_data_types.h]: Pls initialize local varaibles.

**Review by lujiashun — APPROVED** [review]
LGTM

File: storage/tianmu/types/tianmu_data_types.h [hunk 2]
```diff
@@ -490,6 +498,21 @@ static int inline CollationStrCmp(DTCollation coll, const BString &s1, const BSt
                                          s2.len_, 0);
 }
 
+static int inline CollationRealCmp(DTCollation coll, const BString &s1, const BString &s2) {
+  int not_used = 0;
+  char *end_not_used = nullptr;
+  double d1 = coll.collation->cset->strntod(coll.collation, (char *)s1.GetDataBytesPointer(), s1.len_, &end_not_used,
```
Comment (RingsC) [inline storage/tianmu/types/tianmu_data_types.h]: using c++ style type cast.

Comment (RingsC) [inline storage/tianmu/types/tianmu_data_types.h]: pls, ref to `my_strntod`.

File: storage/tianmu/types/tianmu_data_types.h [hunk 3]
```diff
@@ -490,6 +498,21 @@ static int inline CollationStrCmp(DTCollation coll, const BString &s1, const BSt
...
+                                            &not_used);
+
+  double d2 = coll.collation->cset->strntod(coll.collation, (char *)s2.GetDataBytesPointer(), s2.len_, &end_not_used,
+                                            &not_used);
+  if (d1 < d2)
```
Comment (RingsC) [inline storage/tianmu/types/tianmu_data_types.h]: is right? 
```
return (d1 == d2) ? 0 : ( (d1 <d2) ? -1 : 1) ;
```

Comment (RingsC) [inline storage/tianmu/types/tianmu_data_types.h]: is right? 
```
return (d1 == d2) ? 0 : ( (d1 <d2) ? -1 : 1) ;
```

File: storage/tianmu/core/tianmu_attr_exqp.cpp [hunk 4]
```diff
@@ -810,6 +810,18 @@ void TianmuAttr::EvaluatePack_BetweenString_UTF(MIUpdatingIterator &mit, int dim
       return;
     }
   }
+
+  auto check_item_digital = ([&](CQTerm &val) {
+    return ((nullptr != val.item) &&
+            (Item::Type::INT_ITEM == val.item->type() || (Item::Type::DECIMAL_ITEM == val.item->type()) ||
```
Comment (RingsC) [inline storage/tianmu/core/tianmu_attr_exqp.cpp]: `const_val ==  var` is an old fashion, in ok in old compiler, but it's not necessary for a new compiler.

**Review by RingsC — CHANGES_REQUESTED** [review]

**Review by RingsC — APPROVED** [review]

**Pull request closed without merging**